---
title: "waterfallFootprint"
author: "Jordan S Kesner"
date: "October 14, 2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

This code references and adapts original code found here:
https://github.com/EconometricsBySimulation/BivariateSlicer/blob/master/slicedens.R
https://robertgrantstats.wordpress.com/2014/10/08/slice-bivariate-densities-or-the-joy-division-waterfall-plot/
https://en.wikipedia.org/wiki/Kernel_density_estimation
https://stats.stackexchange.com/questions/46302/adding-noise-to-a-matrix-vector
https://stackoverflow.com/questions/31118984/getting-error-while-plotting-contour-map-as-error-in-if-anyh-0-stopban
https://www.rdocumentation.org/packages/MASS/versions/7.3-50/topics/kde2d
http://groups.linguistics.northwestern.edu/speech_comm_group/documents/Presentation_par_cex.pdf
https://www.tydac.ch/color/

Install and load
```{r}
#source("https://bioconductor.org/biocLite.R")
#biocLite("rtracklayer")
#biocLite("BSgenome")
#biocLite("ATACseqQC")
#biocLite("IRanges")
#biocLite("GenomicRanges")
#install.packages("purrr")
#install.packages("repurrrsive")

library(rtracklayer)
library(BSgenome)
library(ATACseqQC)
library(MASS)
library(GenomicRanges)
library(purrr)
library(repurrrsive)
```

slicedens function
```{r}
# x, y, z: data
# slices: number of horizontal slices through the data
# lboost: coefficient to increase the height of the lines
# gboost: coefficient to increase the height of the graph (ylim)
# xinc: horizontal offset for each succesive slice
# (typically something like 1/80)
# yinc: vertical offset for each succesive slice
# bcol: background color
# fcol: fill color for each slice (polygon)
# lcol: line color for each slice
# lwidth: line width
# cutprop: 
# Vary transarency 
# transprop=FALSE
# tmax = .9
# tmin = .2

slicedens<-function(x,y,z=NULL,slices=50,lboost=1,gboost=1,
                    xinc=0,yinc=0.01, bcol="black", fcol="black",
                    lcol="white",lwidth=1, cutprop=FALSE,
                    transprop=FALSE, tmax = .8, tmin = .05,
                    heightprop=FALSE, xlab=FALSE) {
  
  # This function takes a matrix of one or more rgb sets
  # as well as a degree [0,1] and returns a combined
  # color.
  color.mix <- function(cols, degree=0) {
    if (is.null(nrow(cols))) {
      if (class(cols)=="numeric") 
        return(rgb(cols[1],cols[2],cols[3],cols[4]))
      return(cols)
    }
    # Define a function to find elementwise minimum 
    (deg <- degree*(nrow(cols)-1)+1)
    emin <- function(x, y=0) apply(cbind(x, y), 1, min)
    (r <- 1-emin(abs(1:nrow(cols)-deg),1))
    (comb <- apply(cols*r,2,sum))
    mm <- function(x) max(min(x,1),0)
    rgb(mm(comb[1]),
        mm(comb[2]),
        mm(comb[3]),
        mm(comb[4]))
  }
  
  ycut<-min(y)+((0:(slices))*(max(y)-min(y))/slices)
  
  height<-gboost*((slices*yinc)+max(density(x)$y))
  
  plot( c(min(x)-((max(x)-min(x))/10),max(x)+((max(x)-min(x))/10)),
        c(0,height),
        xaxt="n",yaxt="n",ylab="",xlab="")
  rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col=bcol)
  
  # Calcuate the 'degree' for each z value which will be used to
  # choose the color of each slice.
  if (length(z)==length(y)) {
    zmin <- min(z)
    zmax <- max(z)
    zrange <- max(zmax-zmin)
  }
  
  # Define ifcol and ilfol for later reference.
  # Unless noted otherwise, degree=0
  # Meaning the first color will be selected from
  # the rgb color matrix.
  ifcol <- fcol; ilcol <- lcol;  zdeg <- 0
  
  # Define the degree which is the color degree that each slice will
  # contain
  if (length(z)==length(y)){
    meanz <- NULL
    for(i in slices:1)
      meanz[i]<- mean(z[y>=min(ycut[i]) & y<max(ycut[i+1])])
    zdegree<-(meanz-min(meanz, na.rm=TRUE))/
      (max(meanz, na.rm=TRUE)-min(meanz, na.rm=TRUE))
  }
    
  # Loop through and plot slices
  for(i in slices:1) {
    miny<-ycut[i]
    maxy<-ycut[i+1]
    

    gx<-(i-1)*(max(x)-min(x))*xinc
    
    if (cutprop) {
      slLength <- slices*sum(y>=miny & y<maxy)/length(y)
      if (i==slices) gy <- (i-1)*(height)*yinc
      if (i<slices)  gy <- gyLast-(height)*yinc*slLength
      gyLast <- gy
    }
    else gy<-(i-1)*(height)*yinc
    
    if (transprop) {
      trange <- tmax-tmin
      if (is.null(nrow(ifcol)))
        ifcol[4] <- min(trange*slices*sum(y>=miny & y<maxy)/length(y)+tmin,tmax)
      if (!is.null(nrow(ifcol)))
        ifcol[,4] <- min(trange*slices*sum(y>=miny & y<maxy)/length(y)+tmin,tmax)
    }
      
    
    
    # If z is an input vector then grab the color degree from it
    if (length(z)==length(y)) zdeg<-zdegree[i]
    
    # Added the try because distributions without defined upper
    # and lower bounds can give the density function trouble.
    try({
      # Use the color.mixer function to select a color
      fcol<-color.mix(ifcol, zdeg);
      lcol<-color.mix(ilcol, zdeg);
      # Calculte density curves and plot them
      dd<-density(x[y>=miny & y<maxy]);
      if (heightprop) vscale <- lboost*slices*sum(y>=miny & y<maxy)/length(y)
      if (!heightprop) vscale <- lboost
      polygon(dd$x+gx,vscale*dd$y+gy,col=fcol, border=fcol);
      lines(dd$x+gx,vscale*dd$y+gy,col=lcol,lwd=lwidth);
    })
  }
}

```


Change dynamic variables here
```{r}

## Import the records file from the run you want to analyze
load("C:/Users/Jordan/Desktop/h508 wt 01 data/h508/H508_WT_01_records.Rdata")

## Or enter manually the filepaths of signals to analyze
load("C:/Users/Jordan/Desktop/h508 wt 01 data/h508/sigs/GMEB2_motif5_wg_signals.RData")
load("C:/Users/Jordan/Desktop/h508 wt 01 data/h508/sigs/ESRRA_motif4_wg_signals.RData")
load("C:/Users/Jordan/Desktop/h508 wt 01 data/h508/sigs/HNF4A_motif7_wg_signals.RData")

## Transfer the signal objects to a list
signals <- list(H508_WT_01_ESRRA_motif4_wg_signals, H508_WT_01_GMEB2_motif5_wg_signals, H508_WT_01_HNF4A_motif7_wg_signals)

#### Parameters for graph ####
parameters <- list()
parameters$fcol <- c(0.125, 0.529, 0.839,.35) # fill color of slice
parameters$lcol <- c(0,0,0,1) # outline color of slice
parameters$bcol <- c(0,0,0,1) # backgroun color of plot
parameters$lwid <- 1 # line width
parameters$slice <- 10 # number of slices through data
parameters$lboost <- 5 # lboost: coefficient to increase the height of the lines
parameters$gboost <- 1 # gboost: coefficient to increase the height of the graph (ylim)
parameters$xinc <- 1/80 # xinc: horizontal offset for each succesive slice
parameters$yinc <- 1/80 # yinc: verticl offset for each succesive slice
#### end ####


```


waterfallFootprint function
```{r}

# signals = a list of all the signal objects to input

waterfallFootprint <- function(signals, parameters){

num <- length(signals)
slice <- num

##
for (a in 1:num){assign(paste0("signal",a), signals[[a]])} # Transfer the signals objects to new numbered variables
## Collapse each of the signal matrices into a bp insertion probability
for (b in 1:num){
  
# get the plus and minus signal matrix from the genome signals vector
command <- paste0("plus <- signal", a, "[['signal']][['+']]")
eval(parse(text = command, keep.source = getOption("keep.source")))
command <- paste0("minus <- signal", a, "[['signal']][['-']]")
eval(parse(text = command, keep.source = getOption("keep.source")))


# combine the plus and minus signal
signal <- matrix(data=NA,ncol=214,nrow=883672)
signal[,1:214] <- plus[,1:214] + minus[,1:214]

# collapse signals into a single 212 vector
y <- c()

for (a in 1:214){
  y[a] <- sum(signal[,a])
}
  
  
  
  
  
}
  
## 


  
  
} # End waterfallFootprint function
  
  
  
  



```


Load and prepare example ATACseq data
```{r}

# load the sample data
load("~/atacseq/ASCL2_motif1_genome_signals.RData")
signals <- ASCL2_motif1_genome_signals
rm(ASCL2_motif1_genome_signals)
items <- length(signals[["signal"]][["+"]][,1])

# get the quantiles
qnt <- quantile(signals[["bindingSites"]]@elementMetadata@listData[["score"]], na.rm=TRUE)
plot((signals[["bindingSites"]]@elementMetadata@listData[["score"]]))


## get index to split into qunatiles
# first quantile
first <- c(which(signals[["bindingSites"]]@elementMetadata@listData[["score"]] >= qnt[1] & signals[["bindingSites"]]@elementMetadata@listData[["score"]] < qnt[2]))
# second quantile
second <- c(which(signals[["bindingSites"]]@elementMetadata@listData[["score"]] >= qnt[2] & signals[["bindingSites"]]@elementMetadata@listData[["score"]] < qnt[3]))
# third quantile
third <- c(which(signals[["bindingSites"]]@elementMetadata@listData[["score"]] >= qnt[3] & signals[["bindingSites"]]@elementMetadata@listData[["score"]] < qnt[4]))
# fourth quantile
fourth <- c(which(signals[["bindingSites"]]@elementMetadata@listData[["score"]] >= qnt[4] & signals[["bindingSites"]]@elementMetadata@listData[["score"]] <= qnt[5]))





  

  
  signals_first_plus <- list()
  signals_first_minus <- list()
  
  signals_first_plus <- signals[["signal"]][["+"]][first,]
  signals_first_minus <- signals[["signal"]][["-"]][first,]
  
  signals_second_plus <- list()
  signals_second_minus <- list()
  
  signals_second_plus <- signals[["signal"]][["+"]][second,]
  signals_second_minus <- signals[["signal"]][["-"]][second,]
  
  # combine the plus and minus signal
  signal_first <- matrix(data=NA,ncol=214,nrow=679390)
  signal_first[,1:214] <- signals_first_plus[,1:214] + signals_first_minus[,1:214]
  rm(signals_first_plus, signals_first_minus)
  
  
  signal_second <- matrix(data=NA,ncol=214,nrow=204282)
  signal_second[,1:214] <- signals_second_plus[,1:214] + signals_second_minus[,1:214]
  rm(signals_second_plus, signals_second_minus)

  rm(minus, plus, signal, signals)
  
  

# get the plus and minus signal matrix from the genome signals vector
plus <- signals[["signal"]][["+"]]
minus <- signals[["signal"]][["-"]]
# combine the plus and minus signal
signal <- matrix(data=NA,ncol=214,nrow=883672)
signal[,1:214] <- plus[,1:214] + minus[,1:214]

# collapse signals into a single 212 vector
y <- c()

for (a in 1:214){
  y[a] <- sum(signal[,a])
}




## set x and y values
# x = signal at bp position
# y = motif site
x <- rep(1:214,times=679390)

y <- c(signal_first)


```

Make graphs with ATACseq data
```{r}

## Set color of graph slices
# fcol = fill color for each slice
# fcol can either be a color or rgb vector
# lcol is the line color
# bcol is backgroun color
# use rbg(red, green, blue, alpha) function to check color
# check or set colors with https://www.tydac.ch/color/



slicedens(x, y, slices=slice, yinc=yinc, xinc=xinc, lboost=lboost, lwidth=lwid,
          fcol=fcol, bcol=bcol, lcol=lcol, gboost=gboost)


```


```{r}

## Make the plots ##


#### MASS density mapping ####

# cex = magnify text in labels 
# par = used to allow multiple plots in one figure

#n <- 500000; y<-rnorm(n); x<-3*rnorm(n)+y^2
plot(x,y, cex =.2, col=rgb(0,0,0,.1))

# Figure 1

par(mar=c(.1,.1,.1,.1))
par(mfrow = c(2,2))

# Plot out the distribution using MASS density mapping

den3d <- kde2d(x, y)
persp(den3d, box=FALSE, expand=.6)
persp(den3d, box=FALSE, expand=.6, theta=-90)
persp(den3d, box=FALSE, expand=.6, theta=180)
persp(den3d, box=FALSE, expand=.6, theta=90)

#### end ####


#### Side angle version ####

slicedens(x, y, slices=slice, yinc = yinc, lboost=lboost,
          fcol=fcol, bcol=bcol, lcol=lcol,
          gboost=1)


#### end ####


#### Cuts proportionally distanced ####
slicedens(x,y, cutprop=T,
          fcol=fcol, bcol=bcol, lcol=lcol,
          gboost=1)


#### end ####



#### Weighted by transparency ####

slicedens(x,y, transprop=T,
          fcol=fcol, bcol=bcol, lcol=lcol,
          gboost=1)


#### end ####

#### Cuts weighted by height ####


slicedens(x,y, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.3)

slicedens(y,x, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=3)

slicedens(-x,-y, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.3)

slicedens(-y,-x, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=4)

#### end ####


#### With three variables ####

z <- -(abs(x)+abs(y))+rnorm(n)*3
fcol <- rbind(c(0,.1,.5,.5), c(.3,.8,.8,.5), c(1,1,0,.5))
lcol <- rbind(c(0,.3,.3,.8), c(.1,.1,.2,.7), c(0,0,1,.65))


slicedens(x,y,z,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(y,x,z, 
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

slicedens(z,y,x,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(z,x,y,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

#### end ####

#### Stacked ####


# Example:
fcol <- rgb(0,.5,.5,.5)
n <- 500000; y<-rnorm(n); x<-rnorm(n)
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=1.6)
# 
n <- 500000; y<-rnorm(n); x<-rnorm(n)+y*.5
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=1.6)

n <- 500000; y<-rnorm(n); x<-rnorm(n)-y*.5
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=1.6)

n <- 500000; y<-x<-rnorm(n)
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=15)

#### end ####

```




** Full original example code (do not modify) **
```{r eval=FALSE}
# x, y, z: data
# slices: number of horizontal slices through the data
# lboost: coefficient to increase the height of the lines
# gboost: coefficient to increase the height of the graph (ylim)
# xinc: horizontal offset for each succesive slice
# (typically something like 1/80)
# yinc: vertical offset for each succesive slice
# bcol: background color
# fcol: fill color for each slice (polygon)
# lcol: line color for each slice
# lwidth: line width
# cutprop: 
# Vary transarency 
# transprop=FALSE
# tmax = .9
# tmin = .2

slicedens<-function(x,y,z=NULL,slices=50,lboost=1,gboost=1,
                    xinc=0,yinc=0.01, bcol="black", fcol="black",
                    lcol="white",lwidth=1, cutprop=FALSE,
                    transprop=FALSE, tmax = .8, tmin = .05,
                    heightprop=FALSE, xlab=FALSE) {
  
  # This function takes a matrix of one or more rgb sets
  # as well as a degree [0,1] and returns a combined
  # color.
  color.mix <- function(cols, degree=0) {
    if (is.null(nrow(cols))) {
      if (class(cols)=="numeric") 
        return(rgb(cols[1],cols[2],cols[3],cols[4]))
      return(cols)
    }
    # Define a function to find elementwise minimum 
    (deg <- degree*(nrow(cols)-1)+1)
    emin <- function(x, y=0) apply(cbind(x, y), 1, min)
    (r <- 1-emin(abs(1:nrow(cols)-deg),1))
    (comb <- apply(cols*r,2,sum))
    mm <- function(x) max(min(x,1),0)
    rgb(mm(comb[1]),
        mm(comb[2]),
        mm(comb[3]),
        mm(comb[4]))
  }
  
  ycut<-min(y)+((0:(slices))*(max(y)-min(y))/slices)
  
  height<-gboost*((slices*yinc)+max(density(x)$y))
  
  plot( c(min(x)-((max(x)-min(x))/10),max(x)+((max(x)-min(x))/10)),
        c(0,height),
        xaxt="n",yaxt="n",ylab="",xlab="")
  rect(par("usr")[1],par("usr")[3],par("usr")[2],par("usr")[4],col=bcol)
  
  # Calcuate the 'degree' for each z value which will be used to
  # choose the color of each slice.
  if (length(z)==length(y)) {
    zmin <- min(z)
    zmax <- max(z)
    zrange <- max(zmax-zmin)
  }
  
  # Define ifcol and ilfol for later reference.
  # Unless noted otherwise, degree=0
  # Meaning the first color will be selected from
  # the rgb color matrix.
  ifcol <- fcol; ilcol <- lcol;  zdeg <- 0
  
  # Define the degree which is the color degree that each slice will
  # contain
  if (length(z)==length(y)){
    meanz <- NULL
    for(i in slices:1)
      meanz[i]<- mean(z[y>=min(ycut[i]) & y<max(ycut[i+1])])
    zdegree<-(meanz-min(meanz, na.rm=TRUE))/
      (max(meanz, na.rm=TRUE)-min(meanz, na.rm=TRUE))
  }
    
  # Loop through and plot slices
  for(i in slices:1) {
    miny<-ycut[i]
    maxy<-ycut[i+1]
    

    gx<-(i-1)*(max(x)-min(x))*xinc
    
    if (cutprop) {
      slLength <- slices*sum(y>=miny & y<maxy)/length(y)
      if (i==slices) gy <- (i-1)*(height)*yinc
      if (i<slices)  gy <- gyLast-(height)*yinc*slLength
      gyLast <- gy
    }
    else gy<-(i-1)*(height)*yinc
    
    if (transprop) {
      trange <- tmax-tmin
      if (is.null(nrow(ifcol))) 
        ifcol[4] <- min(trange*slices*sum(y>=miny & y<maxy)/length(y)+tmin,tmax)
      if (!is.null(nrow(ifcol))) 
        ifcol[,4] <- min(trange*slices*sum(y>=miny & y<maxy)/length(y)+tmin,tmax)
    }
      
    
    
    # If z is an input vector then grab the color degree from it
    if (length(z)==length(y)) zdeg<-zdegree[i]
    
    # Added the try because distributions without defined upper
    # and lower bounds can give the density function trouble.
    try({
      # Use the color.mixer function to select a color
      fcol<-color.mix(ifcol, zdeg);
      lcol<-color.mix(ilcol, zdeg);
      # Calculte density curves and plot them
      dd<-density(x[y>=miny & y<maxy]);
      if (heightprop) vscale <- lboost*slices*sum(y>=miny & y<maxy)/length(y)
      if (!heightprop) vscale <- lboost
      polygon(dd$x+gx,vscale*dd$y+gy,col=fcol, border=fcol);
      lines(dd$x+gx,vscale*dd$y+gy,col=lcol,lwd=lwidth);
    })
  }
}

n <- 500000; y<-rnorm(n); x<-3*rnorm(n)+y^2
plot(x,y, cex =.2, col=rgb(0,0,0,.1))

# Figure 1

par(mar=c(.1,.1,.1,.1))
par(mfrow = c(2,2))

# Plot out the distribution using MASS density mapping
library(MASS)
den3d <- kde2d(x, y)
persp(den3d, box=FALSE, expand=.6)
persp(den3d, box=FALSE, expand=.6, theta=-90)
persp(den3d, box=FALSE, expand=.6, theta=180)
persp(den3d, box=FALSE, expand=.6, theta=90)

# Figure 2

# fcol can either be a color or rgb vector
fcol <- c(.6,0,0,.35)
lcol <- c(.7,1,1,.35)

slicedens(x,y,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(y,x,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

slicedens(-x,-y,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(-y,-x,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

# Figure 3

# Proportionally distanced cuts
slicedens(x,y, cutprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(y,x, cutprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

slicedens(-x,-y, cutprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(-y,-x, cutprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

# Figure 4

# Transparency weighted cuts
slicedens(x,y, transprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(y,x, transprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

slicedens(-x,-y, transprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(-y,-x, transprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

# Figure 5

# Height weighted cuts
fcol <- c(.6,0,0,.2)
lcol <- c(.7,1,1,.4)

slicedens(x,y, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.3)

slicedens(y,x, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=3)

slicedens(-x,-y, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.3)

slicedens(-y,-x, heightprop=T,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=4)

# Figure 6

# Three variables
z <- -(abs(x)+abs(y))+rnorm(n)*3
fcol <- rbind(c(0,.1,.5,.5), c(.3,.8,.8,.5), c(1,1,0,.5))
lcol <- rbind(c(0,.3,.3,.8), c(.1,.1,.2,.7), c(0,0,1,.65))


slicedens(x,y,z,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(y,x,z, 
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1.6)

slicedens(z,y,x,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

slicedens(z,x,y,
          fcol=fcol, bcol='white', lcol=lcol,
          gboost=1)

# Figure 8

# Example:
fcol <- rgb(0,.5,.5,.5)
n <- 500000; y<-rnorm(n); x<-rnorm(n)
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=1.6)
# 
n <- 500000; y<-rnorm(n); x<-rnorm(n)+y*.5
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=1.6)

n <- 500000; y<-rnorm(n); x<-rnorm(n)-y*.5
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=1.6)

n <- 500000; y<-x<-rnorm(n)
slicedens(x,y,
          fcol=fcol, bcol='white', lcol=rgb(0,0,.75,.5),
          gboost=15)
          
# Figure 7

# I downloaded the IPUMS data from https://www.ipums.org/
# I selected age wage income and person weight. 
# You will need to download your own data if you want to do this part.
# The challenge with the data is that it is in fixed width 
# format so you may need to customize this read.fwf command.
ipums <- read.fwf("C:/Data/usa_00028.dat/usa_00028.dat", c(10,3,6))
head(ipums)
names(ipums) <- c('pwght', 'age', 'wginc')

ipums <- ipums[ipums$wginc!=999999 & ipums$wginc!=0,]
plot(density(log(ipums$wginc[ipums$age==16])))

par(mar=c(.1,.1,.1,.1))
par(mfrow = c(2,2))


fcol <- c(.3,.8,.8,.1); lcol <- c(.5,1,1,.2)

slicedens(log(ipums$wginc), ipums$age,
          fcol=fcol, bcol='black', lcol=lcol,
          gboost=.8)

slicedens(log(ipums$wginc), -ipums$age,
          fcol=fcol, bcol='black', lcol=lcol,
          gboost=.8)

slicedens(ipums$age, log(ipums$wginc),
          fcol=fcol, bcol='black', lcol=lcol,
          gboost=.2)

slicedens(log(ipums$wginc), ipums$age, heightprop=TRUE,
          fcol=fcol, bcol='black', lcol=lcol,
          gboost=2)

# Figure 9
```





---
title: "ridgelineFootprintPlots"
author: "Jordan S. Kesner"
date: "October 12, 2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

Install
```{r}
install.packages("ggplot2")
install.packages("ggridges")
install.packages("iridis")
```

Load
```{r}
library(ggplot2)
library(ggridges)
library(viridis)
```

Examples
```{r}
ggplot(diamonds, aes(x = price, y = cut)) +
  geom_density_ridges(scale = 4) + theme_ridges() +
  scale_y_discrete(expand = c(0.01, 0)) +   # will generally have to set the `expand` option
  scale_x_continuous(expand = c(0, 0))      # for both axes to remove unneeded padding
#> Picking joint bandwidth of 458


data <- data.frame(x = 1:5, y = rep(1, 5), height = c(0, 1, 3, 4, 2))
ggplot(data, aes(x, y, height = height)) + geom_ridgeline()

d <- data.frame(x = rep(1:5, 3), y = c(rep(0, 5), rep(1, 5), rep(2, 5)),
                height = c(0, 1, 3, 4, 0, 1, 2, 3, 5, 4, 0, 5, 4, 4, 1))
ggplot(d, aes(x, y, height = height, group = y)) + 
  geom_ridgeline(fill = "lightblue")


# density plots
ggplot(iris, aes(x = Sepal.Length, y = Species)) + geom_density_ridges()

ggplot(iris, aes(x = Sepal.Length, y = Species)) + geom_density_ridges2()

# The grouping aesthetic does not need to be provided if a categorical variable is mapped onto the y axis, but it does need to be provided if the variable is numerical.
# modified dataset that represents species as a number
iris_num <- transform(iris, Species_num = as.numeric(Species))

# does not work, causes error
# ggplot(iris_num, aes(x = Sepal.Length, y = Species)) + geom_density_ridges()

# works 
ggplot(iris_num, aes(x = Sepal.Length, y = Species_num, group = Species_num)) + 
  geom_density_ridges()


```

Functions
```{r}

pullMotif <- function(signals, number){

# get plus and minus strand signals
plus <- signals[["signal"]][["+"]]
minus <- signals[["signal"]][["-"]]
# get the motif matching score
score <- score <- TCF7_motif1_genome_signals[["bindingSites"]]@elementMetadata@listData[["score"]]

# declare a matrix for combines values
combined <- matrix(data = NA, length(plus[,1]), 212)
# set the row names
rownames(combined) <- rownames(plus)

# populate the combined list with plus and minus signals
for (a in 1:length(plus[,1]))
  {
  combined[a,1:212] <- plus[a,1:212] + minus[a,1:212]
  }

# get the total signal
totals <- apply(combined, 1, sum)

# sort them to get highest ten
sorted <- sort(totals, decreasing = TRUE)
# get the rownames of those ten
named <- names(sorted)
# subset the combined matrix based on row names
newlist <- combined[named,]

motifs_df <- data.frame(bp = rep(1:212,number), group = rep(1:number,each = 212))

index <- 1
  
  for (j in 1:number)
  {
    for (k in 1:212)
    {
    motifs_df[index,"signal"] <- newlist[((number+1)-j),k]
    index <- index + 1
    
    } # end for (b in 1:212)
  } # end for (a in 1:numer)

  return(motifs_df)
}

```


Template code for ridges plots
```{r}

## make a ridges plot with one motif
one_motif <- pullMotif(TCF7_motif1_genome_signals, 1)
#one_motif_log <- one_motif
#one_motif_log[,3] <- log(one_motif_log[3], 2)

one_motif_z <- one_motif
one_motif_z[,3] <- scale(one_motif_z[,3], center = FALSE, scale = TRUE) 


ggplot(one_motif_z, aes(bp, signal, height = signal, group = group, color = "black")) + 
  geom_ridgeline(fill = "lightblue")

## make a ridges plot with two motifs
two_motif <- pullMotif(TCF7_motif1_genome_signals, 2)

ggplot(two_motif, aes(bp, signal, height = signal, group = group, color = "black")) + 
  geom_ridgeline(fill = "lightblue")

## make a ridges plot with ten motifs
ten_motif <- pullMotif(TCF7_motif1_genome_signals, 10)

ggplot(ten_motif, aes(bp, signal, height = signal, group = group)) + 
  geom_ridgeline(fill = "lightblue")


```


Testing code
```{r}


## try a density plot
ggplot(one_motif, aes(x = signal, y = group)) + geom_density_ridges()

# This works
ggplot(ten_motif, aes(x = signal, y = group, group = group)) + 
  geom_density_ridges()



```




Rank sites by matching score, plot by signal inside motif
```{r}

# TCF7 binding motif is 12 bp
sigs <- TCF7_motif1_genome_signals
sites <- sigs[["bindingSites"]]

num <- length(sigs[["signal"]][["+"]][,1])


# get plus and minus
plus <- sigs[["signal"]][["+"]]
minus <- sigs[["signal"]][["-"]]
score <- score <- TCF7_motif1_genome_signals[["bindingSites"]]@elementMetadata@listData[["score"]]


# declare a matrix for combines values
combined <- matrix(data = NA, length(plus[,1]), 212)
# set the row names
rownames(combined) <- rownames(plus)


# populate the combined list with plus and minus signals
for (a in 1:length(plus[,1]))
  {
  combined[a,1:212] <- plus[a,1:212] + minus[a,1:212]
  }



# make matrix for the motif signal
total_motif_signal <- matrix(data = NA, length(plus[,1]), 3)
# calc and transfer the signal
for (a in 1:length(plus[,1]))
  {
  total_motif_signal[a,1] <- sum(combined[a,100:112])
  total_motif_signal[a,2] <- score[a]
  }


# mark groups for the signals
for (a in 1:length(plus[,1]))
  {
  if (total_motif_signal[a,2] > 0.9921911){total_motif_signal[a,3] <- 1}
  if (total_motif_signal[a,2] <= 0.9921911){total_motif_signal[a,3] <- 2}
  }

# make into a dataframe
motifs_df <- data.frame(signal = total_motif_signal[,1], score = total_motif_signal[,2], group = total_motif_signal[,3])


# plot distributions
# This works
ggplot(motifs_df, aes(x = signal, y = score, group = score)) + 
  geom_density_ridges(bandwidth = 0.5)


```


Draw a ridgeline plot where x = bp position, y = motif score, height = signal
```{r}


# TCF7 binding motif is 12 bp
sigs <- TCF7_motif1_genome_signals
sites <- sigs[["bindingSites"]]

num <- length(sigs[["signal"]][["+"]][,1])


# get plus and minus
plus <- sigs[["signal"]][["+"]]
minus <- sigs[["signal"]][["-"]]
score <- score <- TCF7_motif1_genome_signals[["bindingSites"]]@elementMetadata@listData[["score"]]


# declare a matrix for combines values
combined <- matrix(data = NA, length(plus[,1]), 212)
# set the row names
rownames(combined) <- rownames(plus)


# populate the combined list with plus and minus signals
for (a in 1:length(plus[,1]))
  {
  combined[a,1:212] <- plus[a,1:212] + minus[a,1:212]
  }

# get the total signal
totals <- apply(combined, 1, sum)

# sort them to get highest ten
sorted <- sort(totals, decreasing = TRUE)
# get the rownames of those ten
named <- names(sorted)
# subset the combined matrix based on row names
newlist <- combined[named,]

index <- 1
# make a matrix
number <- 2
motifs <- matrix(data = NA, (number * 212), )
data.frame(bp = rep(1:212,number), group = rep(1:number,each = 212))

for (a in 1:number){
  for (b in 1:212){
    motifs_df["signal"][index] <- newlist[b,a]
    index <- index + 1
  }
}


# make matrix for the motif signal
total_motif_signal <- matrix(data = NA, length(plus[,1]), 3)
# calc and transfer the signal
for (a in 1:length(plus[,1]))
  {
  total_motif_signal[a,1] <- sum(combined[a,100:112])
  total_motif_signal[a,2] <- score[a]
  }


# mark groups for the signals
for (a in 1:length(plus[,1]))
  {
  if (total_motif_signal[a,2] > 0.9921911){total_motif_signal[a,3] <- 1}
   if (total_motif_signal[a,2] <= 0.9921911){total_motif_signal[a,3] <- 2}
  }

# make into a dataframe
motifs_df <- data.frame(signal = total_motif_signal[,1], score = total_motif_signal[,2], group = total_motif_signal[,3])



## make a ridges plot with two motifs
two_motif <- pullMotif(TCF7_motif1_genome_signals, 2)

ggplot(two_motif, aes(x, y, height = height, group = y)) + 
  geom_ridgeline(fill = "lightblue")



```





